### PR TITLE
Feature/startup typescript

### PR DESCRIPTION
**Pull request by zhouyuhe**
chart component changed to typescript
Also changed socketService into typescript

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,15 +17,23 @@
     "redux": "^4.0.4",
     "redux-logger": "^3.0.6",
     "socket.io": "^2.2.0",
-    "socket.io-client": "^2.2.0"
+    "socket.io-client": "^2.2.0",
+    "typescript": "^3.6.4"
```
Comment (Sharakai) [inline package.json]: `typescript` should be in `devDependencies`

File: server/handleConnection.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ exports.handleConnection = socket => {
     Object.values(timerIDs).forEach(clearInterval);
     timerIDs.stockTicker = callAndStartIntervals(
       emitStockTicker,
-      5000,
+      ONE_DAY_IN_MS,
```
Comment (Sharakai) [inline server/handleConnection.js]: Q: This looks wrong. You're now only emitting the stock price once every day?

File: src/features/chart/components/Chart.tsx [hunk 3]
```diff
@@ -53,25 +53,28 @@ const formatDate = (isoDate, ChartRange) => {
...
+    (state: AppState) => state.stockTickerData
   );
-  const { selectedStockTicker } = useSelector(state => state.stockTickerData);
-  const updateChartRange = stock => dispatch(updateChartAction(stock));
-  const onClickHandler = event => {
-    updateChartRange(event.target.value);
-  };
+  const updateChartRange = (stock: string) =>
+    dispatch(updateChartAction(stock));
+  // const onClickHandler: MouseEventHandler<HTMLButtonElement> = (event) => {
```
Comment (Sharakai) [inline src/features/chart/components/Chart.tsx]: nit: Please remove commented code if it's no longer used.

File: src/features/chart/components/Chart.tsx [hunk 4]
```diff
@@ -53,25 +53,28 @@ const formatDate = (isoDate, ChartRange) => {
...
-    selectedChartData.map(data => ({
-      date: formatDate(data.date, selectedChartRange),
-      close: data.close
-    }));
+  const chartData = selectedChartData
+    ? selectedChartData.map(data => ({
+        date: formatDate(data.date, selectedChartRange),
+        close: data.close
+      }))
+    : [];
```
Comment (Sharakai) [inline src/features/chart/components/Chart.tsx]: Note: This is a perfectly fine way of writing this, but you can also write it as:

```ts
const chartData = (selectedChartData || []).map(data => ({
  date: formatDate(data.date, selectedChartRange),
  close: data.close
}));
```

File: src/features/chart/components/Chart.tsx [hunk 5]
```diff
@@ -80,7 +83,8 @@ export const Chart = () => {
               <button
                 key={label}
                 className={`chart__button ${activeClass}`}
-                onClick={onClickHandler}
+                onClick={() => updateChartRange(value)}
+                // onClick={onClickHandler}
```
Comment (Sharakai) [inline src/features/chart/components/Chart.tsx]: nit: Please remove commented code

File: src/features/chart/components/Chart.tsx [hunk 6]
```diff
@@ -125,11 +129,13 @@ export const Chart = () => {
...
-              value: String(selectedStockTicker.latestPrice),
-              position: "right",
-              fill: "#e95656"
-            }}
+            label={
+              {
+                value: String(selectedStockTicker.latestPrice),
+                position: "right",
+                fill: "#e95656"
+              } as any
```
Comment (Sharakai) [inline src/features/chart/components/Chart.tsx]: You shouldn't ever have to cast `as any` unless it's absolutely required.
In this case, the typings for `<ReferenceLine />` are telling you that you've made a mistake, and using `as any` is ignoring the problem.

Fix:
```ts
label={
  <Label value={selectedStockTicker.latestPrice} position="right" fill="#e95656" />
}
```

File: src/features/chart/redux/actions.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { UPDATE_CHART_RANGE, UPDATE_CHART_DATA } from "./constants";
+
+export type UpdateChartRangeAction = {
```
Comment (Sharakai) [inline src/features/chart/redux/actions.ts]: I highly recommend making a generic Action type, and re-using that for all these actions. I believe @keithbloom has been showing you how to type these, but the most basic is:

```ts
type Action<Type extends string> = { type: Type };
type ActionWithPayload<Type extends string, Payload> = Action<Type> & { payload: Payload };
```

This `UpdateChartRangeAction` would then change to:
```ts
type UpdateChartRangeAction = ActionWithPayload<typeof UPDATE_CHART_RANGE, string>;
```

File: src/features/chart/redux/constants.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export const UPDATE_CHART_DATA: "UPDATE_CHART_DATA" = "UPDATE_CHART_DATA";
```
Comment (Sharakai) [inline src/features/chart/redux/constants.ts]: You don't need to define a in-line type for the constant, as Typescript can infer it from its assigned value:

```suggestion
export const UPDATE_CHART_DATA = "UPDATE_CHART_DATA";
```

File: src/features/chart/redux/constants.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
...
+export const UPDATE_CHART_RANGE: "UPDATE_CHART_RANGE" = "UPDATE_CHART_RANGE";
```
Comment (Sharakai) [inline src/features/chart/redux/constants.ts]:
```suggestion
export const UPDATE_CHART_RANGE = "UPDATE_CHART_RANGE";
```

File: src/features/chart/redux/constants.ts [hunk 10]
```diff
@@ -0,0 +1,3 @@
...
+export const RESET: "RESET" = "RESET";
```
Comment (Sharakai) [inline src/features/chart/redux/constants.ts]:
```suggestion
export const RESET = "RESET";
```

File: src/features/chart/redux/middleware.ts [hunk 11]
```diff
@@ -1,8 +1,19 @@
...
+import { SocketService } from "../../../services/socketService";
 
-export const chartMiddleware = ({
+export type Dependencies = {
+  socketService: SocketService;
+};
+
+export type ChartMiddleware = (
+  dependencies: Dependencies
+) => Middleware<{}, AppState, Dispatch<AnyAction>>;
```
Comment (Sharakai) [inline src/features/chart/redux/middleware.ts]: nit: You don't need the 3rd generic parameter here. It is defaulted to `Dispatch`
```suggestion
) => Middleware<{}, AppState>;
```

File: src/store/reducer.ts [hunk 12]
```diff
@@ -8,6 +8,8 @@ import { chartReducer } from "../features/chart";
 import { headlineReducer } from "../features/headline";
 import { stockReducer } from "../reducer";
 
+export type AppState = ReturnType<typeof combinedReducer>;
```
Comment (Sharakai) [inline src/store/reducer.ts:11]: 👍 Nice.

**Review by Sharakai — DISMISSED** [review]

File: server/handleConnection.js (hunk 2 above)
Comment (zhouyuhe) [reply server/handleConnection.js]: I want to look at my data structure in the redux store so I will change it back later on.

File: src/features/chart/components/Chart.tsx (hunk 4 above)
Comment (zhouyuhe) [reply src/features/chart/components/Chart.tsx]: This looks nicer

File: server/handleConnection.js (hunk 2 above)
Comment (Sharakai) [reply server/handleConnection.js]: As discussed in person, this change should **never** be pushed through into `development`, as it isn't safe for production. You'd also be affecting every other developer on this project.

If you want to make this change whilst developing, then ensure you revert it before PR'ing.
In your case, I recommend not changing it at all, and using the features of `redux-devtools` instead.

Comment (zhouyuhe) [reply server/handleConnection.js]: Got it, will change it back

File: src/features/chart/components/Chart.tsx [hunk 13]
```diff
@@ -25,9 +28,9 @@ const tenors = [
   { value: "max", label: "MAX" }
 ];
 
-const yaxisFormat = item => item.toFixed(2);
+const yaxisFormat = (item: number) => item.toFixed(2);
 
-const formatDate = (isoDate, ChartRange) => {
+const formatDate = (isoDate: Date, ChartRange: string) => {
```
Comment (steinalex) [inline src/features/chart/components/Chart.tsx]: q: Isn't `isoDate` a string that's sent from the server, which is then converted into a `Date`?

File: src/features/chart/redux/middleware.ts [hunk 14]
```diff
@@ -16,7 +27,7 @@ export const chartMiddleware = ({
   }
   if (action.type === BOOTSTRAP) {
     const socket = socketService.get();
-    socket.on("chartData", payload => {
+    socket.on("chartData", (payload: { date: Date; close: number }[]) => {
```
Comment (steinalex) [inline src/features/chart/redux/middleware.ts]: q: As mentioned before, this is a `string` that's sent from the server

File: src/features/chart/redux/actions.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+export const updateChartAction = (
+  chartRange: string
+): UpdateChartRangeAction => ({
+  type: UPDATE_CHART_RANGE,
+  payload: chartRange
+});
+
+export type UpdateChartDataAction = ActionWithPayload<
+  typeof UPDATE_CHART_DATA,
+  { date: Date; close: number }[] | null
```
Comment (steinalex) [inline src/features/chart/redux/actions.ts]: q: Maybe it'll be better to create an `interface` or `type` which contains your data shape.

You'll then be able to pass this down to your `Middleware.js` file, so your `socket.on` function will become:
`socket.on("chartData", (payload: ChartData[] => {`

File: src/features/chart/components/Chart.tsx (hunk 13 above)
Comment (Sharakai) [reply src/features/chart/components/Chart.tsx]: Very good spot!

**Comment by steinalex** [comment]
You've also changed your socketService to `TypeScript`, so might be worth mentioning this in your description

**Review by steinalex — APPROVED** [review]

**Pull request merged**